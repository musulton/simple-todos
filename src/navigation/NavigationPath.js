export const SPLASH_PATH = 'SPLASH';
export const LOGIN_PATH = 'LOGIN';
export const TODO_PATH = 'TODO';
