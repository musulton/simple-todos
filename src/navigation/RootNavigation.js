import {createNavigationContainerRef, StackActions} from '@react-navigation/native';

export const navigationRef = createNavigationContainerRef()

export function navigate(name, params, isReplace = false) {
    if (navigationRef.isReady()) {
        if (isReplace) {
            navigationRef.current.dispatch(StackActions.replace(name))
        } else {
            navigationRef.navigate(name, params);
        }

    }
}
