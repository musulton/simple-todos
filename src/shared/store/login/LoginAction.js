import {LOGIN, LOGOUT} from "../../utils/constants";

export function login(loginInfo) {
    return {
        type: LOGIN,
        payload: loginInfo
    }
}

export function logout() {
    return {
        type: LOGOUT,
    }
}
