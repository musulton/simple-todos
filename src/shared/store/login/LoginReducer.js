import {LOGIN, LOGOUT} from "../../utils/constants";

const initialState = {
    isLoggedIn: false,
    error: ''
};

const LoginReducer = (state = initialState, action) => {
    switch (action.type) {
        case LOGIN: {
            return {
                error: '',
                isLoggedIn: action.payload
            }
        }
        case LOGOUT: {
            return {
                error: '',
                isLoggedIn: false
            };
        }
        default:
            return state;
    }
}
export default LoginReducer;
