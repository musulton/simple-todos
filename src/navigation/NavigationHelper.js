import * as RootNavigation from './RootNavigation';

export function goToScreen(routePath, isReplace) {
    RootNavigation.navigate(routePath, null, isReplace);
}
