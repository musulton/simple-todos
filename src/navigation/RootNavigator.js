import React from "react";
import {NavigationContainer} from "@react-navigation/native";
import {createNativeStackNavigator} from "@react-navigation/native-stack";

import LoginScreen from "../screens/LoginScreen/LoginScreen"
import ToDoScreen from "../screens/TodoScreen/ToDoScreen";
import SplashScreen from "../screens/SplashScreen/SplashScreen";
import {LOGIN_PATH, SPLASH_PATH, TODO_PATH} from "./NavigationPath";
import PopupMenu from "../shared/components/PopupMenu";
import LoginService from "../services/LoginService";
import {Login} from "../screens/LoginScreen/Login";
import {navigationRef} from "./RootNavigation";

const Stack = createNativeStackNavigator();

const RootNavigator = () => {
    return <NavigationContainer ref={navigationRef}>
        <Stack.Navigator initialRouteName={SPLASH_PATH}>
            <Stack.Screen name={SPLASH_PATH} component={SplashScreen} options={{headerShown: false}}/>
            <Stack.Screen name={LOGIN_PATH} options={{headerShown: false}}>
                {props => <LoginScreen {...props} login={() => Login(LoginService)} />}
            </Stack.Screen>
            <Stack.Group
                screenOptions={({navigation}) => ({
                    headerStyle: {
                        backgroundColor: '#f2f2f2',
                    },
                    headerShadowVisible: false,
                    headerRight: () => <PopupMenu navigation={navigation}/>
                })}
            >
                <Stack.Screen name={TODO_PATH} component={ToDoScreen} options={{ title: '' }}/>
            </Stack.Group>
        </Stack.Navigator>
    </NavigationContainer>
}

export default RootNavigator;
