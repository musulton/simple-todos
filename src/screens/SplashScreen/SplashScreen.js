import React, {useEffect} from 'react';
import {Image, StyleSheet, View,} from 'react-native';
import {LOGIN_PATH, TODO_PATH} from "../../navigation/NavigationPath";
import {enigma} from "../../shared/assets/images";
import LocalStorage from "../../shared/utils/LocalStorage";
import {goToScreen} from "../../navigation/NavigationHelper";
import {login} from "../../shared/store/login/LoginAction";
import {useDispatch} from "react-redux";

const SplashScreen = () => {
    const dispatch = useDispatch();

        useEffect(() => {
            setTimeout(async () => {
                const isLoggedIn = !!await LocalStorage().getData('token')
                const toRouteName = isLoggedIn ? TODO_PATH : LOGIN_PATH;
                dispatch(login({isLoggedIn}));
                goToScreen(toRouteName, true)
            }, 3000)
        })

        return (
            <View style={styles.content}>
                <Image source={enigma}/>
            </View>
        );
    }
;

const styles = StyleSheet.create({
    content: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    }
});

export default SplashScreen;
