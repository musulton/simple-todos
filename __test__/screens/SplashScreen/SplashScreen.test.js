import {render, screen, act} from "@testing-library/react-native";

import SplashScreen from "../../../src/screens/SplashScreen/SplashScreen";
import {login} from "../../../src/shared/store/login/LoginAction";
import {goToScreen} from "../../../src/navigation/NavigationHelper";
import {LOGIN_PATH, TODO_PATH} from "../../../src/navigation/NavigationPath";

const mockGetData = jest.fn()

jest
    .mock("react-redux", () => ({
        ...jest.requireActual("react-redux"),
        useDispatch: jest.fn(() => jest.fn())
    }))
    .mock("../../../src/shared/assets/images", () => ({
        enigma: "dummy/path/enigma.jpg"
    }))
    .mock("../../../src/shared/store/login/LoginAction", () => ({
        login: jest.fn()
    }))
    .mock("../../../src/navigation/NavigationHelper", () => ({
        goToScreen: jest.fn()
    }))
    .mock("../../../src/shared/utils/LocalStorage", () => ({
        __esModule: true,
        default: jest.fn(() => ({
            getData: mockGetData
        }))
    }))

describe("SplashScreen", () => {
    beforeAll(() => {
        jest.useFakeTimers();
    })

    afterAll(() => {
        jest.clearAllMocks()
    })

    test("should return correctly", () => {
        render(<SplashScreen />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    test("should navigate to TODO_PATH when has token", async () => {
        mockGetData.mockImplementation(() => true)

        render(<SplashScreen />)

        await act(() => {
            jest.runOnlyPendingTimers()
        })

        expect(login).toHaveBeenCalledWith({isLoggedIn: true})
        expect(goToScreen).toHaveBeenCalledWith(TODO_PATH, true)
    })

    test("should navigate to LOGIN_PATH when has not token", async () => {
        mockGetData.mockImplementation(() => false)

        render(<SplashScreen />)

        await act(() => {
            jest.runOnlyPendingTimers()
        })

        expect(login).toHaveBeenCalledWith({isLoggedIn: false})
        expect(goToScreen).toHaveBeenCalledWith(LOGIN_PATH, true)
    })
})
