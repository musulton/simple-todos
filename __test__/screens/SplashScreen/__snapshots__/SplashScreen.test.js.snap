// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplashScreen should return correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    source="dummy/path/enigma.jpg"
  />
</View>
`;
